Dip/Dip/Definition.swift

```diff
@@ -50,20 +50,62 @@ func ==(lhs: DefinitionKey, rhs: DefinitionKey) -> Bool {
 public enum ComponentScope {
   /// Indicates that a new instance of the component will be created each time it's resolved.
   case Prototype
+  /// Indicates that instances will be reused during resolve but will be discurded when topmost `resolve` method returns.
+  case ObjectGraph
   /// Indicates that resolved component should be retained by container and always reused.
   case Singleton
 }
 
 ///Definition of type T describes how instances of this type should be created when they are resolved by container.
 public final class DefinitionOf<T>: Definition {
```

Comment:
We can try this:
- make `DefinitionOf<T, F>` where F will be type of factory and use it for factory property instead of `Any`
- add container argument to `resolveDependencies` method of `DefinitionOf` and use it to reinsert updated definition
- add `remove<T, F>(_: DefinitionOf<T, F>)` and `register<T, F>(_: DefinitionOf<T, F>)` to container (these methods can be handy anyway)

If that would not work (I suspect there will be ambiguity in method calls caused by generics) I would stick to using `final class`, adding fatal error and removing `return self` in `resolveDependencies`.